sql-serializer/src/main/java/io/cdap/pipeline/sql/serializer/bigquery/BigQuerySQLSerializer.java

```diff
@@ -0,0 +1,247 @@
+      if (t.hasProject()) {
+        b.append(t.getProjectName());
+        b.append('.');
+      }
+      if (t.hasDatabase()) {
+        b.append(t.getDatabaseName());
+        b.append('.');
+      }
+      b.append(t.getTableName());
+      b.append('`');
+      if (t.hasAlias() && defineAliasIfExists) {
+        b.append(" AS ");
+        b.append(t.getAlias());
+      }
+    } else {
+      b.append('`');
+      b.append(t.getAlias());
+      b.append('`');
+    }
+    return b;
+  }
+
+  private StringBuilder appendQueryableSQL(StringBuilder b, Queryable s) {
+    QueryableType type = s.getSourceType();
+    if (type.equals(QueryableType.QUERY)) {
+      return appendQuerySQL(b, (StructuredQuery) s, true);
+    } else if (type.equals(QueryableType.TABLE)) {
+      return appendTableSQL(b, (Table) s, true);
+    } else {
+      throw new IllegalArgumentException("Invalid readable source type");
+    }
+  }
+
+  private StringBuilder appendFilterSQL(StringBuilder b, Filter f, Set<Column> definedColumns) {
+    b.append('(');
+    if (f.isLeaf()) {
+      appendOperandSQL(b, f.getLeftOperand(), definedColumns);
+      b.append(' ');
+      appendPredicateOperatorSQL(b, f.getOperation());
+      b.append(' ');
+      appendOperandSQL(b, f.getRightOperand(), definedColumns);
+    } else {
+      appendFilterSQL(b, f.getLeftFilter(), definedColumns);
+      b.append(' ');
+      appendPredicateOperatorSQL(b, f.getOperation());
+      b.append(' ');
+      appendFilterSQL(b, f.getRightFilter(), definedColumns);
+    }
+    b.append(')');
+    return b;
+  }
+
+  private StringBuilder appendOperandSQL(StringBuilder b, Operand o, Set<Column> definedColumns) {
+    OperandType type = o.getOperandType();
+    if (type.equals(OperandType.COLUMN)) {
+      Column col = (Column) o;
+      // Only use the alias form if the column was not defined the in current nesting level
+      return appendColumnSQL(b, col, false, !definedColumns.contains(col));
+    } else if (type.equals(OperandType.CONSTANT)) {
+      return appendConstantSQL(b, (Constant) o);
+    } else {
+      throw new IllegalArgumentException("Invalid operand type");
+    }
+  }
+
+  private StringBuilder appendPredicateOperatorSQL(StringBuilder b, PredicateOperatorType operation) {
+    switch(operation) {
+      case OR:
+        b.append("OR");
+        break;
+      case AND:
+        b.append("AND");
+        break;
+      case LESS:
+        b.append("<");
+        break;
+      case LESS_OR_EQUAL:
+        b.append("<=");
+        break;
+      case EQUAL:
+        b.append("=");
+        break;
+      case GREATER:
+        b.append(">");
+        break;
+      case GREATER_OR_EQUAL:
+        b.append(">=");
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid predicate operator type");
+    }
+    return b;
+  }
+
+  private StringBuilder appendConstantSQL(StringBuilder b, Constant o) {
+    switch (o.getConstantType()) {
+      case DATETIME:
```

Comment:
are we adding more types later?

Reply:
Possibly, I can imagine adding booleans or decimals later